### PR TITLE
Add request promise resolver

### LINKED ISSUE #2: React Router v4
Instead of a Provider type class that exposes registerDependency we could use the new `staticContext`:

```js
static contextTypes = {
  router: PropTypes.shape({
    staticContext: PropTypes.object,
  }).isRequired,
}

componentWillMount() {
  const { staticContext } = this.context.router

  if (staticContext) {
    staticContext.registerDependency(this.dispatchIfNeeded(this.props))
  }
}
```

### PR DESCRIPTION
**Pull request by simenbrekken**
This adds a new module `resolver` which is capable of resolving promises recursively without an external context.

Usage example:

```js
const renderApplication = () => renderToString(<Application />)

try {
  const markup = await resolveRequestPromises(renderApplication, { maxIterations: 2 })

  res.send(markup)
} catch (error) {
  next(error)
}
```

Closes #2 and #3

This still needs to be documented properly along with examples for which I've created a separate issue #5

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,18 +1,13 @@
...
-export {
-  getActionRequestOptions,
-  getRequestState,
- } from './utils'
-
-// FIXME: Redux DevTools extension needs to be configured using serializedAction
-// for symbols to show up in actions as JSON.stringify strips symbols
-export const FETCH = 'fetch' // Symbol('fetch')
+export { getActionRequestOptions, getRequestState } from './utils'
+export { registerRequestPromise, resolveRequestPromises } from './resolver'
```
Comment (einarlove) [inline src/index.js]: Why export registerRequestPromise? What is the use-case?

File: src/resolver.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+  if (iterations < maxIterations && promises.length > 0) {
+    await Promise.all(promises)
+    performIteration(resolver, iterations + 1, maxIterations)
+  }
+
+  return Promise.resolve(result)
+}
+
+export const registerRequestPromise = createRequestPromise => {
```
Comment (einarlove) [inline src/resolver.js]: Why not
```js
registerRequestPromise = promise => {
  if (resolving && promise) {
    promises.push(promise)
  }
}
// fetchData.js
componentWillMount() {
  registerRequestPromise(this.dispatchIfNeeded(this.props))
}
```

**Review by einarlove — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 1 above)
Comment (simenbrekken) [reply src/index.js]: Agreed, removed in f1b52fd

File: src/resolver.js (hunk 2 above)
Comment (simenbrekken) [reply src/resolver.js]: Because the promise shouldn't be created unless we are actually resolving.still create a promise even if it shouldn't

**Review by einarlove — APPROVED** [review]

**Pull request merged**